### PR TITLE
[FIX] account: relax constraint-change UoM on product template

### PR DESCRIPTION
**Pull request by yosa-odoo**
This PR introduced the constraint on the change of UoM on a product template.

Restricting on the uom_id is a bit too restrictive. Instead, the restriction will be made on the uom category

opw-3039419

### CODE REVIEW

File: addons/account/models/product.py [hunk 1]
```diff
@@ -57,9 +57,9 @@ def get_product_accounts(self, fiscal_pos=None):
     @api.constrains('uom_id')
     def _check_uom_not_in_invoice(self):
         for template in self:
-            invoices = self.env['account.move.line'].sudo().search([('product_id.product_tmpl_id.id', '=', template.id)], limit=1)
-            if invoices:
-                raise ValidationError(_('The product "%s" is used in invoices. You cannot change its Unit of Measure.', template.display_name))
+            aml = self.env['account.move.line'].sudo().search([('product_id.product_tmpl_id.id', '=', template.id)], limit=1)
+            if aml and aml.product_uom_id.category_id != self.uom_id.category_id:
+                raise ValidationError(_("You cannot change the unit of measure as there are already move lines with this product. If you want to change the unit of measure, you should rather archive this product and create a new one."))
```
Comment (epictete) [inline addons/account/models/product.py:62]: Why not include all your conditions in the search domain directly (and use the `template` instead of `self` :wink:) ?
I would rephrase the message a bit to make slightly more "diplomatic" while using a less technical term (Journal Entries vs. move lines).

Something like:
```diff
diff --git a/addons/account/models/product.py b/addons/account/models/product.py
index eca1f0833916..89533f53462f 100644
--- a/addons/account/models/product.py
+++ b/addons/account/models/product.py
@@ -57,9 +57,14 @@ class ProductTemplate(models.Model):
     @api.constrains('uom_id')
     def _check_uom_not_in_invoice(self):
         for template in self:
-            aml = self.env['account.move.line'].sudo().search([('product_id.product_tmpl_id.id', '=', template.id)], limit=1)
-            if aml and aml.product_uom_id.category_id != self.uom_id.category_id:
-                raise ValidationError(_("You cannot change the unit of measure as there are already move lines with this product. If you want to change the unit of measure, you should rather archive this product and create a new one."))
+            aml_domain = [
+                ('product_id.product_tmpl_id.id', '=', template.id),
+                ('product_uom_id.category_id.id', '!=', template.uom_id.category_id.id),
+            ]
+            if self.env['account.move.line'].sudo().search(aml_domain, limit=1):
+                raise ValidationError(
+                    _("This product's unit of measure is already being used in posted Journal Entries.\n"
+                      "If you want to change it, please archive this product and create a new one."))
```

**Review by epictete — COMMENTED** [review]
Litle details :slightly_smiling_face:

File: addons/account/models/product.py (hunk 1 above)
Comment (yosa-odoo) [reply addons/account/models/product.py:62]: Yes, I like the domain use :)

For the message wouldn't it be ok to say maybe:

"This product is already being used in posted Journal Entries.\n"
"If you want to change its Unit of Measure, please archive this product and create a new one."
?

So we insist on the product instead of the uom (that can be used by other products)

Comment (epictete) [reply addons/account/models/product.py:62]: Sure, it was just a suggestion, it does not have to be the exact same phrase :wink:

**Comment by epictete** [comment]
@robodoo r+
